Signal/src/ViewControllers/MessagesViewController.m

```diff
@@ -2598,53 +2602,55 @@ - (void)yapDatabaseModified:(NSNotification *)notification {
-                  YapCollectionKey *collectionKey = rowChange.collectionKey;
-                  if (collectionKey.key) {
-                      [self.messageAdapterCache removeObjectForKey:collectionKey.key];
-                  }
-                  [self.collectionView reloadItemsAtIndexPaths:@[ rowChange.indexPath ]];
-                  break;
-              }
-          }
-      }
+        for (YapDatabaseViewRowChange *rowChange in messageRowChanges) {
+            switch (rowChange.type) {
+                case YapDatabaseViewChangeDelete: {
+                    [self.collectionView deleteItemsAtIndexPaths:@[ rowChange.indexPath ]];
+
+                    YapCollectionKey *collectionKey = rowChange.collectionKey;
+                    if (collectionKey.key) {
+                        [self.messageAdapterCache removeObjectForKey:collectionKey.key];
+                    }
+
+                    break;
+                }
+                case YapDatabaseViewChangeInsert: {
+                    [self.collectionView insertItemsAtIndexPaths:@[ rowChange.newIndexPath ]];
+                    scrollToBottom = YES;
+
+                    TSInteraction *interaction = [self interactionAtIndexPath:rowChange.newIndexPath];
+                    if (![interaction isKindOfClass:[TSOutgoingMessage class]]) {
+                        shouldAnimateScrollToBottom = YES;
+                    }
+                    break;
+                }
+                case YapDatabaseViewChangeMove: {
+                    [self.collectionView deleteItemsAtIndexPaths:@[ rowChange.indexPath ]];
+                    [self.collectionView insertItemsAtIndexPaths:@[ rowChange.newIndexPath ]];
+                    break;
+                }
+                case YapDatabaseViewChangeUpdate: {
+                    YapCollectionKey *collectionKey = rowChange.collectionKey;
+                    if (collectionKey.key) {
+                        [self.messageAdapterCache removeObjectForKey:collectionKey.key];
+                    }
+                    [self.collectionView reloadItemsAtIndexPaths:@[ rowChange.indexPath ]];
+                    break;
+                }
+            }
+        }
     }
         completion:^(BOOL success) {
-          if (!success) {
-              [self.collectionView.collectionViewLayout
-                  invalidateLayoutWithContext:[JSQMessagesCollectionViewFlowLayoutInvalidationContext context]];
-              [self.collectionView reloadData];
-          }
-          if (scrollToBottom) {
-              [self scrollToBottomAnimated:shouldAnimateScrollToBottom];
-          }
+            OWSAssert([NSThread isMainThread]);
+
+            if (!success) {
+                [self.collectionView.collectionViewLayout
+                    invalidateLayoutWithContext:[JSQMessagesCollectionViewFlowLayoutInvalidationContext context]];
+                [self.collectionView reloadData];
+            }
+            if (scrollToBottom && self.userIsScrolling) {
```

Comment:
I'm not sure how you got here, but want to double check, because I'd assume you'd want this condition inverted.

 e.g. you don't want to scroll *while* the user is *already* scrolling.